### PR TITLE
Multibind backfills

### PR DESCRIPTION
**Pull request by shellderp**

### CODE REVIEW

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/BackfilaStartupConfigurator.kt [hunk 1]
```diff
@@ -24,7 +24,7 @@ internal class BackfilaStartupConfigurator @Inject internal constructor(
   private val config: BackfilaClientConfig,
   private val backfilaClient: BackfilaClient,
   @ForBackfila private val moshi: Moshi,
-  @ForBackfila private val backfills: Map<String, KClass<out Backfill<*, *>>>
+  @ForBackfila private val backfills: MutableMap<String, KClass<out Backfill<*, *>>>
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/BackfilaStartupConfigurator.kt:27]: has to be mutable map for mapbinding to work

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt [hunk 2]
```diff
@@ -44,5 +51,21 @@ class BackfilaModule(
...
 }
 
+private fun mapBinder(binder: Binder) = MapBinder.newMapBinder(
+    binder,
+    object : TypeLiteral<String>() {},
+    object : TypeLiteral<KClass<out Backfill<*, *>>>() {},
+    ForBackfila::class.java
+)
+
+inline fun <reified T : Backfill<*, *>> Binder.bindBackfill() {
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt:61]: is there a smarter way to do this?

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt [hunk 3]
```diff
@@ -48,10 +47,14 @@ internal class ClientMiskTestingModule(
...
-    ))
+    install(
+        BackfilaModule(
+            BackfilaClientConfig(
+                url = "test.url", slack_channel = "#test"
+            )
+        )
+    )
+
+    binder().bindBackfill<TestBackfill>()
```
Comment (shellderp) [inline client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt:58]: eg

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt (hunk 2 above)
Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt:61]: I'm no expert but I think that's right.

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/BackfilaStartupConfigurator.kt (hunk 1 above)
Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/BackfilaStartupConfigurator.kt:27]: Do we want a point at which it is no longer mutable?

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt (hunk 3 above)
Comment (mpawliszyn) [reply client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt:58]: Yeah this reified version looks good. 

We probably want to log what gets registered. I think this already happens on the backfila side though.

Comment (mpawliszyn) [inline client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt:58]: Is this the right place for this though?
Really I only want this to register for the SinglePartition use case.

We want to specify the backfills to register. Either by passing in the classes or adding a function that does the binding.

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt (hunk 2 above)
Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt:61]: Maybe @swankjesse has a suggestion but I believe this is correct.

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt (hunk 3 above)
Comment (shellderp) [reply client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt:58]: I think it is the right place and we would just have multiple backfill classes for different tests, wdyt?

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/BackfilaStartupConfigurator.kt (hunk 1 above)
Comment (shellderp) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/BackfilaStartupConfigurator.kt:27]: the two cases we have are used in very small scopes so I'm not worried

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt [hunk 4]
```diff
@@ -4,37 +4,44 @@ import app.cash.backfila.client.misk.client.BackfilaClientConfig
...
-    for (backfill in backfills) {
-      map[backfill.jvmName] = backfill
+    if (backfills == null) {
+      mapBinder(binder())
+    } else {
+      val map = mutableMapOf<String, KClass<out Backfill<*, *>>>()
+      for (backfill in backfills) {
+        map[backfill.jvmName] = backfill
+      }
+      bind(object : TypeLiteral<MutableMap<String, KClass<out Backfill<*, *>>>>() {})
```
Comment (swankjesse) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt:41]: why not multibind them here? Then you can gradually transition from non-multibindings to multibindings

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt (hunk 2 above)
Comment (swankjesse) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt:61]: I prefer a slightly different syntax where we let the application itself do the multibind() call.

Here’s how we document it in StaticResourceEntry:
```
/**
 * This data class is used with Guice multibindings. Register instances by calling `multibind()`
 * in a `KAbstractModule`:
 *
 * ```
 * multibind<StaticResourceEntry>().toInstance(StaticResourceEntry(...))
 * ```
 */
class StaticResourceEntry(
  url_path_prefix: String = "/",
  private val resourcePath: String
) : ValidWebEntry(url_path_prefix = url_path_prefix) {
 ...
}
```

Comment (swankjesse) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt:61]: Or perhaps what we do for Web Actions?
```
    install(WebActionModule.create<InternalErrorAction>())
```

Comment (swankjesse) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt:61]: (this way you don’t need to subclass a special module class, which I find a bit awkward)

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt (hunk 4 above)
Comment (shellderp) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt:41]: doh, fixed

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt (hunk 2 above)
Comment (shellderp) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaModule.kt:61]: thanks, that's way better

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt (hunk 3 above)
Comment (shellderp) [reply client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt:58]: done

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/hibernate/SinglePartitionHibernateBackfillTest.kt [hunk 5]
```diff
@@ -1,64 +1,27 @@
...
 import misk.hibernate.Id
-import misk.hibernate.Query
 import misk.hibernate.Session
 import misk.hibernate.Transacter
 import okhttp3.internal.toImmutableList
-import okio.ByteString
 import okio.ByteString.Companion.encodeUtf8
 import org.assertj.core.api.Assertions.assertThat
 import org.junit.jupiter.api.Test
+import javax.inject.Inject
```
Comment (shellderp) [inline client-misk/src/test/kotlin/app/cash/backfila/client/misk/hibernate/SinglePartitionHibernateBackfillTest.kt:17]: if someone knows how to fix intellij doing this..

**Pull request merged**